### ISSUE TITLE
Incorrect port used on Kubernetes

### ISSUE THREAD

**Issue by drmaciej** [codecentric/spring-boot-admin#1771]
In k8s, it looks like `management.port` (as mentioned in https://codecentric.github.io/spring-boot-admin/current/#_converting_serviceinstances) is not used. I'm explicitly registering an instance of `KubernetesServiceInstanceConverter`, but it doesn't seem to do the trick.

On my application "details" page I can see that `port.management` is set to 9999. I can also see that the Pod's IP is correct and the port next to it is correct as well. The call to the `health` endpoint is made using 9999. Every other call is made using port 80.

I can see that links in responses from `/actuator` do not contain the port name, they look like `http://my-app/actuator/env`, and I suspect this is the endpoint that is used (including the implicit port 80), and `management.port` is not used.

Is there a way I can force boot-admin to always make all the requests on the right port?

**Comment by Na3blis** [comment]
This sounds like the same issue I'm experiencing, https://github.com/codecentric/spring-boot-admin/issues/1667

**Comment by drmaciej** [comment]
It does look like the same issue, or at least a very similar one.

For completeness, I should add that my Boot Admin is 2.4.0 (via starter server), my Boot version is 2.4.6.

**Comment by Glosur** [comment]
Are you using the Spring Boot Admin Client or any Discovery Client ?

**Comment by drmaciej** [comment]
I actually tried with both and experienced the same problem with both.

I ended up suppressing the `EndpointDetector` with a reduced one, so that endpoints are not discovered or parsed, but provided via properties:
```
    @Primary
    @Bean
    public EndpointDetector endpointDetector(InstanceRepository instanceRepository, InstanceWebClient.Builder instanceWebClientBuilder) {
        InstanceWebClient instanceWebClient = instanceWebClientBuilder.build();
        ProbeEndpointsStrategy strategy = new ProbeEndpointsStrategy(instanceWebClient, adminServer.getProbedEndpoints());
        return new EndpointDetector(instanceRepository, strategy);
    }
```

This finally resulted in the right port being used.

**Comment by Glosur** [comment]
Could you provide the return of the /actuator endpoint ?

Are the links correct in the return ?

**Comment by drmaciej** [comment]
The links are correct as far as I can tell. There's no port in them. The calls are made directly to a Pod, but I have certain requirements and certain workarounds in place to make the calls flow via my Service. 
This is why when the call gets to the container the port is no longer present in the request URL, and Boot builds the links without the port in them. This happens in `WebMvcLinksHandler#links`, with `linksResolver.resolveLinks(request.getRequestURL().toString())`

Interestingly enough, the calls to `/health` are always made on the right port, like mentioned above.

My output:
```
{
  "_links": {
    "self": {
      "href": "http://my-service/actuator",
      "templated": false
    },
    "beans": {
      "href": "http://my-service/actuator/beans",
      "templated": false
    },
    "caches-cache": {
      "href": "http://my-service/actuator/caches/{cache}",
      "templated": true
    },
    "caches": {
      "href": "http://my-service/actuator/caches",
      "templated": false
    },
    "health": {
      "href": "http://my-service/actuator/health",
      "templated": false
    },
    "health-path": {
      "href": "http://my-service/actuator/health/{*path}",
      "templated": true
    },
    "info": {
      "href": "http://my-service/actuator/info",
      "templated": false
    },
    "conditions": {
      "href": "http://my-service/actuator/conditions",
      "templated": false
    },
    "configprops": {
      "href": "http://my-service/actuator/configprops",
      "templated": false
    },
    "env": {
      "href": "http://my-service/actuator/env",
      "templated": false
    },
    "env-toMatch": {
      "href": "http://my-service/actuator/env/{toMatch}",
      "templated": true
    },
    "logfile": {
      "href": "http://my-service/actuator/logfile",
      "templated": false
    },
    "loggers": {
      "href": "http://my-service/actuator/loggers",
      "templated": false
    },
    "loggers-name": {
      "href": "http://my-service/actuator/loggers/{name}",
      "templated": true
    },
    "heapdump": {
      "href": "http://my-service/actuator/heapdump",
      "templated": false
    },
    "threaddump": {
      "href": "http://my-service/actuator/threaddump",
      "templated": false
    },
    "prometheus": {
      "href": "http://my-service/actuator/prometheus",
      "templated": false
    },
    "metrics-requiredMetricName": {
      "href": "http://my-service/actuator/metrics/{requiredMetricName}",
      "templated": true
    },
    "metrics": {
      "href": "http://my-service/actuator/metrics",
      "templated": false
    },
    "scheduledtasks": {
      "href": "http://my-service/actuator/scheduledtasks",
      "templated": false
    },
    "mappings": {
      "href": "http://my-service/actuator/mappings",
      "templated": false
    },
    "refresh": {
      "href": "http://my-service/actuator/refresh",
      "templated": false
    },
    "features": {
      "href": "http://my-service/actuator/features",
      "templated": false
    },
    "serviceregistry": {
      "href": "http://my-service/actuator/serviceregistry",
      "templated": false
    },
    "jolokia": {
      "href": "http://my-service/actuator/jolokia",
      "templated": false
    }
  }
}
```

**Comment by Na3blis** [comment]
>         ProbeEndpointsStrategy strategy = new ProbeEndpointsStrategy(instanceWebClient, adminServer.getProbedEndpoints());

@drmaciej what does your adminServer ProbedEndpoints property look like? Unfortunately I've just been using SBA to only tell me whether the service is up or down, but haven't been able to use all the other features because of this issue. 

@Glosur My services are all on port 8080, with no other port (#1667).

**Comment by drmaciej** [comment]
@Na3blis I took what was hardcoded in boot-admin's code and massaged it a little, which then resulted in
```
      probed-endpoints:
        - "threaddump:dump"
        - auditevents
        - beans
        - caches
        - conditions
        - configprops
        - env
        - flyway
        - health
        - heapdump
        - httptrace
        - "httptrace:trace"
        - info
        - jolokia
        - logfile
        - loggers
        - mappings
        - metrics
        - quartz
        - refresh
        - scheduledtasks
        - sessions
        - threaddump
 ```

@Glosur multiple Pods are not a problem. I made a few changes both in my k8s setup (mostly pertaining to how Istio routes traffic in my clusters) and in boot-admin to make it all work.  I do have 2 ports on my Service (an "application" one and a "management" one). I had one port before, and switching to two didn't make a difference. I tried multiple ways of pointing boot-admin at the right port, but to no avail. I ended up switching from server-side discovery to client self-registration when trying to fix the problem (which didn't help much anyway, but I dislike the discovery mechanism, so I'm glad I no longer need to use it). 

Long story short, no matter what I did, I could not force boot-admin to use the right port. The above is the only way I managed to do it.

I think the takeaway here is that boot-admin needs to be told (with an additional property?) to use the port from config above all, and that should suppress what it finds using its own discovery mechanisms.

**Comment by Na3blis** [comment]
Oh, ok. I thought you had somehow added in the ports to that list or something. So just by using doing that it resolved the issue of it not having the port on each endpoint? I'll give it a try and see if it works for me too, thanks

**Comment by drmaciej** [comment]
good news!

**Comment by ulischulte** [comment]
Since the problem seems to be solved, I'm closing this issue.

**Comment by odu14ick** [comment]
Hey @Na3blis , @drmaciej 
can you explain in a little bit more detail what you did to work around the issue?

I'm having the very same issue you had in https://github.com/codecentric/spring-boot-admin/issues/1667 .

Still not extremely clear for me what needs to be done. Is the Client side discovery a must here? I figure the "which didn't help much anyway" part means it's not relevant.

Or do I just need to put the "probed-endpoints" configuration somwhere? Just a stray devops here trying to help my dev.  SBA is dark woods for me.

**Comment by mvhiller2001** [comment]
This worked like a charm, thank you so much.

**Comment by Na3blis** [comment]
Unfortunately this is still around. I switched to discovery and was hoping I'd be able to remove this custom code so that I wouldn't have to manually updated probedEndpoints anymore, but everything other than health and metadata is broken if you don't manually create the endpointDetector. I will say, with my local kubernetes (k3s) and skaffold, it works with no problems, but as soon as I deployed to GKE, everything but those two endpoints broke without the custom detector

**Comment by erikpetzold** [comment]
@Na3blis it is a bit difficult to help as even you say it is working in some environments but not in others.

In general, the response of the `/actuator` endpoint should contain the correct urls with port.
Besides that, the client forwarding the request to the instance should use the same uri that it used to invoke the `/actuator` endpoint.

You could enable more logs, like
```
    de.codecentric.boot.admin.server.web.client.InstanceExchangeFilterFunctions: trace
    de.codecentric.boot.admin.server.services.endpoints: trace
```
and check if you see where the port is going lost.
